internal/plugin/connectors/tcp/mssql/connector.go

```diff
@@ -159,96 +167,126 @@ func (connector *SingleUseConnector) Connect(
+		driverConn, err := driverConnector.Connect(loginContext)
 
-	// Since the communication between the client and Secretless must be unencrypted,
-	// we fool the client into thinking that it's talking to a server that does not support
-	// encryption.
-	preloginResponse[mssql.PreloginENCRYPTION] = []byte{mssql.EncryptNotSup}
+		connectionResultChan <- connectResult{
+			conn: driverConn.(*mssql.Conn),
+			err:  err,
+		}
+	} ()
 
-	// Write the prelogin packet back to the user
-	err = mssql.WritePreloginWithPacketType(clientBuffer, preloginResponse, mssql.PackReply)
-	if err != nil {
-		wrappedError := errors.Wrap(err, "failed to write prelogin response")
-		connector.sendError(wrappedError)
-		return nil, wrappedError
+	var loginAck mssql.LoginAckStruct
+	// Block continuation until error or driver has completed connection
+responseLoop:	for {
+		select {
+		// preLoginResponse is received from the server
+		case preloginResponse := <- connInterceptor.PreLoginResponse:
+			// Since the communication between the client and Secretless must be unencrypted,
+			// we fool the client into thinking that it's talking to a server that does not support
+			// encryption.
+			preloginResponse[mssql.PreloginENCRYPTION] = []byte{mssql.EncryptNotSup}
+
+			// Write the prelogin packet back to the user
+			err = mssql.WritePreloginResponse(connector.clientBuff, preloginResponse)
+			if err != nil {
+				wrappedError := errors.Wrap(err, "failed to write prelogin response")
+				connector.sendError(wrappedError)
+				return nil, wrappedError
+			}
+
+			// We parse the client's Login packet so that we can pass on params to the server.
+			clientLogin, err := mssql.ReadLogin(connector.clientBuff)
+			if err == io.EOF {
+				return nil, err
+			}
+			if err != nil {
+				wrappedError := errors.Wrap(err, "failed to handle client login")
+				connector.sendError(wrappedError)
+				return nil, wrappedError
+			}
+
+			// Send the client login to the mssql driver which it can use to pass client params
+			// to the server. This channel is set as a value on the context passed to the mssql
+			// driver on construction.
+			connInterceptor.ClientLogin <- *clientLogin
+			// TODO: where this is consumed inside go-mssqldb, it real
+			break
+		// a protocol error is received from the server
+		case protocolErr := <- connInterceptor.ServerError:
+			err = protocolErr
+			break responseLoop
+		// a loginAck is sent from the server, this will be followed by connectionResultChan
+		// unless there's an issue
+		case loginAck = <- connInterceptor.ServerLoginAck:
+			break
+		// connect is finished. this case is visited when there is non-protocol error or
+		// after loginAck
+		case res := <-connectionResultChan:
+			if res.err != nil {
+				err = res.err
+			} else {
+				// Verify the driverConn is an mssql driverConn object and get its
+				// underlying transport
+				connector.backendConn = res.conn.NetConn()
+			}
+
+			break responseLoop
+		}
```

Comment:
I wrote this response to your previous version of this code, but everything still applies, double so now....

---

You must have been waiting to see my reaction on this one.... I will _not_ disappoint you :)

I think 1. the labelled loop is both unnecessary and 2. it obscured what is actual pretty simple logic.

As you note in the comments:

```go
// a loginAck is sent from the server, this will be followed by connectPhaseFinished
// unless there's an issue
```
Thus we should be able to rewrite this as:

1. a single (no loop) select which waits for just the loginAck or error.
2. an if which returns if there was an error
3. a 2nd single select (no loop) which waits for just the `connectPhaseFinished` or another error

This makes the expected sequential actions explicit, gets rid of the loop and breaks and label, and as a bonus will eliminate the `if... else`.